sklearn/preprocessing/tests/test_data.py

```diff
@@ -2136,6 +2149,69 @@ def test_categorical_encoder_ordinal_inverse():
     exp = np.array(X, dtype=object)
     assert_array_equal(enc.inverse_transform(X_tr), exp)
 
+    # incorrect shape raises
+    X_tr = np.array([[0, 1, 1, 2], [1, 0, 1, 0]])
+    msg = re.escape('Shape of the passed X data is not correct')
+    assert_raises_regex(ValueError, msg, enc.inverse_transform, X_tr)
+
+
+def test_categorical_encoder_dtypes():
+    # check that dtypes are preserved when determining categories
+    enc = CategoricalEncoder()
+    exp = np.array([[1., 0., 1., 0.], [0., 1., 0., 1.]], dtype='float64')
+
+    X = np.array([[1, 2], [3, 4]], dtype='int64')
+    enc.fit(X)
+    assert all([enc.categories_[i].dtype == 'int64' for i in range(2)])
+    assert_array_equal(enc.transform(X).toarray(), exp)
+
+    X = np.array([[1, 2], [3, 4]], dtype='float64')
+    enc.fit(X)
+    assert all([enc.categories_[i].dtype == 'float64' for i in range(2)])
+    assert_array_equal(enc.transform(X).toarray(), exp)
+
+    X = np.array([['a', 'b'], ['c', 'd']])  # string dtype
+    enc.fit(X)
+    assert all([enc.categories_[i].dtype == X.dtype for i in range(2)])
+    assert_array_equal(enc.transform(X).toarray(), exp)
+
+    X = np.array([[1, 'a'], [3, 'b']], dtype='object')
+    enc.fit(X)
+    assert all([enc.categories_[i].dtype == 'object' for i in range(2)])
+    assert_array_equal(enc.transform(X).toarray(), exp)
+
+    X = [[1, 2], [3, 4]]
+    enc.fit(X)
+    assert all([np.issubdtype(enc.categories_[i].dtype, np.integer)
+                for i in range(2)])
+    assert_array_equal(enc.transform(X).toarray(), exp)
+
+    X = [[1, 'a'], [3, 'b']]
+    enc.fit(X)
+    assert all([enc.categories_[i].dtype == 'object' for i in range(2)])
+    assert_array_equal(enc.transform(X).toarray(), exp)
+
+
+def test_categorical_encoder_dtypes_pandas():
+    # check dtype (similar to test_categorical_encoder_dtypes for dataframes)
+    try:
+        import pandas as pd
+    except ImportError:
+        raise SkipTest("pandas is not installed")
+
+    enc = CategoricalEncoder()
+    exp = np.array([[1., 0., 1., 0.], [0., 1., 0., 1.]], dtype='float64')
+
+    X = pd.DataFrame({'A': [1, 2], 'B': [3, 4]}, dtype='int64')
+    enc.fit(X)
+    assert all([enc.categories_[i].dtype == 'int64' for i in range(2)])
+    assert_array_equal(enc.transform(X).toarray(), exp)
+
+    X = pd.DataFrame({'A': [1, 2], 'B': ['a', 'b']})
+    enc.fit(X)
+    assert all([enc.categories_[i].dtype == 'object' for i in range(2)])
```

Comment:
Is this the behaviour we want? Or just what happens to happen now...

Reply:
Yes, this is what we want *assuming* a dataframe first is converted to an array (then object is the best we can do to preserve types). But in case of the CategoricalEncoder which works column by column, there is not really a need to convert to an array.